022-feature-gates.md

```diff
@@ -0,0 +1,105 @@
+# Feature Gates
+
+This proposal suggests to add support for Feature Gates in the Strimzi operators.
+Feature Gates should give us additional options how to control and mature different behaviors in the operators.
+
+## Current situation
+
+Currently, any features or behaviors of the operator are controlled by one of these two mechanisms:
+
+* The Strimzi custom resources (as a form of an API)
+* The state of the Kubernetes resources
+
+The following examples show how are these mechanisms used to give a better idea:
+
+* The `strimzi.io/use-connector-resources` annotation in the `KafkaConnect` custom resource defines whether the Connector Operator should be enabled or not for given Kafka Connect resource.
+* The `useServiceDnsDomain` in the listener configuration of the `Kafka` custom resource defines what kind of address should be used for given listener.
+  Whether fully qualified DNS name or not.
+* When the operator during Kafka upgrade sees that the old pods are still using the TLS sidecars, it will automatically loosen the TLS configuration to make sure the upgrade can be completed.
+
+## Motivation
+
+In many cases, the mechanisms described above work fine and are well suited.
+But there are also limitations:
+
+* Changes to the custom resources APIs have permanent character.
+  We can remove fields from the CRDs only when changing the API versions.
+  So adding fields to the Kafka CR is not a good option to gradually roll out changes or change a default behavior over multiple releases.
+* Sometimes, you also want to just change the behavior of the operator in general.
+  But you want to avoid having endless list of `enableFeatureX`, `disableFeatureY`, or `useZ` fields in the custom resources.
```

Comment:
Sorry, I'm not sure I follow this comment. I mean in theory, you can change the default of the field in the API in the same way you change the you change the default of the feature gate. Or do you mean that that would be against best API practices?

Reply:
Changing a default value is often not backwards compatible, because it's often difficult to reason that it couldn't possibly break something for users. Because of this it means that not only do early adopters end up with lots of `enableFeatureX`, `disableFeatureY` it means that anyone who wants to use those features (even when they become _recommended_ things which all Strimzi users should use) ends up with those fields.

But I can see this is rather a convoluted argument, so I'm fine with not adding it to the proposal.

Reply:
Got it. I tried to add it to the text.